exporter/filelogexporter/exporter.go

```diff
@@ -0,0 +1,131 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filelogexporter
+
+import (
+	"context"
+	"github.com/golang/groupcache/lru"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+)
+
+// fileLogExporter is the implementation of file exporter that writes telemetry data to a file
+// in Protobuf-JSON format.
+type fileLogExporter struct {
+	path                 string
+	basedirAttributeKey  string
+	filenameAttributeKey string
+	maxOpenFiles         int
+	files                *lru.Cache
+	mutex                sync.Mutex
+	logger               *zap.Logger
+}
+
+func (e *fileLogExporter) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (e *fileLogExporter) ConsumeLogs(_ context.Context, ld pdata.Logs) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	resourceLogs := ld.ResourceLogs()
+	for i := 0; i < resourceLogs.Len(); i++ {
+		resourceLog := resourceLogs.At(i)
+		var resourceBase string
+		if baseDir, exists := resourceLog.Resource().Attributes().Get(e.basedirAttributeKey); exists {
+			resourceBase = baseDir.AsString()
+		}
+		ills := resourceLog.InstrumentationLibraryLogs()
+		for j := 0; j < ills.Len(); j++ {
+			libraryLog := ills.At(j)
+			lls := libraryLog.Logs()
+			for k := 0; k < lls.Len(); k++ {
+				log := lls.At(k)
+				logBase := resourceBase
+				if baseDir, exists := log.Attributes().Get(e.basedirAttributeKey); exists {
+					logBase = baseDir.AsString()
+				}
+				if fileName, exists := log.Attributes().Get(e.filenameAttributeKey); exists {
```

Comment:
nit: this body is a bit long, can we do early return inverting the condition?

Reply:
Not sure how it will make the body smaller. I can extract it in another method ?

Reply:
I am sorry I did not explain myself correctly. I meant that the body is long and identation makes it harder to read (due to max line length so using the early return will reduce the identation.

Reply:
Done.